cmd/registration-tracker/main.go

```diff
@@ -0,0 +1,377 @@
+			rs.prevRegIds[strSharedKey]++
+			rs.currRegIds[strSharedKey] = 1
+			rs.m.Unlock()
+			return ErrDupReg
+		} else {
+			rs.currRegIds[strSharedKey] = 1
+			rs.m.Unlock()
+			err = rs.AddRegistration(asn, cc, isIPv4, reg)
+			if err != nil {
+				// handle err if needed
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func handleC2SError(logger *log.Logger, IPVerSupport string, err error, reg *cj.DecoyRegistration){
+
+        if errors.Is(err, phantoms.ErrLegacyMissingAddrs) ||
+                errors.Is(err, phantoms.ErrLegacyAddrSelectBug) ||
+                errors.Is(err, phantoms.ErrLegacyV0SelectionBug) ||
+                errors.Is(err, transports.ErrUnknownTransport) {
+                logger.Debugf("DEBUG: unexpected err creating %s registration: %v, %v", IPVerSupport, err, reg)
+        } else {
+                logger.Errorf("unexpected err creating %s registration: %v, %v", IPVerSupport, err, reg)
+        }
+}
+
+func main() {
+	var zmqAddress string
+	flag.StringVar(&zmqAddress, "zmq-address", "ipc://@zmq-proxy", "Address of ZMQ proxy")
+	flag.Parse()
+
+	logger := log.New(os.Stdout, "[REGTRACK] ", golog.Ldate|golog.Lmicroseconds)
+
+	// parse toml station configuration
+	conf, err := cj.ParseConfig()
+	if err != nil {
+		logger.Fatalf("failed to parse app config: %v", err)
+	}
+
+	privkey, err := conf.ParsePrivateKey()
+	if err != nil {
+		logger.Fatalf("error parseing private key: %s", err)
+	}
+
+	geoipDB, err := geoip.New(conf.RegConfig.DBConfig)
+	if err != nil {
+		logger.Fatalf("failed to create geoip database: %v", err)
+	}
+
+	regstats := regStats{
+		logger:          logger,
+		m:               &sync.Mutex{},
+		v4Registrations: make(map[uint]*transportStats),
+		v6Registrations: make(map[uint]*transportStats),
+		currRegIds:      make(map[string]int),
+		prevRegIds:      make(map[string]int),
+		GeoIP:           geoipDB,
+	}
+
+	ctx, _ := context.WithCancel(context.Background())
+	regChan := make(chan interface{}, 10000)
+	zmqIngester, err := cj.NewZMQIngest(zmqAddress, regChan, privkey, conf.ZMQConfig)
+	if err != nil {
+		logger.Fatal("error creating ZMQ Ingest: %w", err)
+	}
+
+	// Receive registration updates from ZMQ Proxy as subscriber
+	go zmqIngester.RunZMQ(ctx)
+
+	regManager := cj.NewRegistrationManager(&cj.RegConfig{})
+
+	// Add supported transport options for registration validation
+	for transportType, transport := range enabledTransports {
+		err = regManager.AddTransport(transportType, transport)
+		if err != nil {
+			logger.Fatalf("failed to add transport: %v", err)
+		}
+	}
+
+	ticker := time.NewTicker(30 * time.Second)
+	go func() {
+		for range ticker.C {
+			regstats.PrintAndReset()
+		}
+	}()
+
+	msgChan := make(chan interface{}, 10000)
+
+	// Read from regChan to which the zmqIngester writes
+	for data := range regChan {
+
+		select {
+		case <- ctx.Done():
+			logger.Fatal("closing all ingest threads")
+		case msgChan <- data:
```

Comment:
Why have regChan return an array of msgChans that you have to read from? Is there a reason to avoid it providing the messages directly?